### PR TITLE
Support new visual editor in Javascript SDK

### PR DESCRIPTION
**Pull request by jdorn**
## Features and Changes

Currently, the visual editor requires adding a standalone `<script>` tag to the page with its own separate context and targeting attributes.

This PR merges the visual editor logic into the Javascript SDK for a better developer experience and richer feature set.

## TODO
- [x] Update types
- [x] Add experiments to the feature-repository
- [x] Helper methods on the GrowthBook class to get/set experiments, the page URL, and DOM mutator instance
- [x] Run and apply visual editor experiments automatically and whenever the URL or targeting attributes change
- [x] Ability to manually trigger an experiment
- [x] URL targeting logic
- [x] DOM mutator package and rollup bundle
- [ ] Automatically detect URL changes in SPAs (if simple)
- [ ] Documentation for the Javascript SDK
- [ ] Documentation for the React SDK
- [ ] Tests for URL targeting
- [ ] Tests for setting experiments
- [ ] Tests for updating experiments when targeting changes
- [ ] Tests for new feature-repository payload
- [ ] Tests for manually triggered experiments

### CODE REVIEW

File: packages/sdk-js/src/util.ts [hunk 1]
```diff
@@ -62,6 +62,42 @@ export function getUrlRegExp(regexString: string): RegExp | undefined {
...
+  let hasIncludeRules = false;
+  let isIncluded = false;
+
+  for (let i = 0; i < targets.length; i++) {
+    const match = _evalURLTarget(url, targets[i].type, targets[i].pattern);
+    if (targets[i].include === false) {
+      if (match) return false;
+    } else {
+      hasIncludeRules = true;
+      if (match) isIncluded = true;
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/util.ts:76]: wouldn't this just be `return true` and then skip the final check for `isIncluded || !hasIncludeRules`

File: packages/sdk-js/src/util.ts [hunk 2]
```diff
@@ -62,6 +62,42 @@ export function getUrlRegExp(regexString: string): RegExp | undefined {
...
+
+  return isIncluded || !hasIncludeRules;
+}
+
+function _evalURLTarget(
+  url: string,
+  type: UrlTargetType,
+  pattern: string
+): boolean {
+  const pathOnly = url.replace(/^https?:\/\//, "").replace(/^[^/]*\//, "/");
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/util.ts:88]: probably fine for now, but do we need to worry about query strings (and their order)?

File: packages/sdk-js/package.json [hunk 3]
```diff
@@ -17,6 +17,7 @@
   "engines": {
     "node": ">=10"
   },
+  "sideEffects": false,
```
Comment (bttf) [inline packages/sdk-js/package.json:20]: I understand this is needed for tree shaking, however ... We do have side effects, right? (Adding properties to the global `window`object).

File: packages/sdk-js/src/util.ts (hunk 1 above)
Comment (bttf) [reply packages/sdk-js/src/util.ts:76]: @bryce-fitzsimons is this blocking?

File: packages/sdk-js/src/util.ts (hunk 2 above)
Comment (bttf) [reply packages/sdk-js/src/util.ts:88]: same question @bryce-fitzsimons

**Pull request merged**